File: GaestehausWild/Model/Geo.swift
import Foundation
import CoreLocation

enum Geo {
    static let guesthouse = CLLocationCoordinate2D(latitude: 49.4224662, longitude: 10.9807374)

    static func distanceInKilometres(latitude: Double, longitude: Double) -> Double {
        let origin = CLLocation(latitude: guesthouse.latitude, longitude: guesthouse.longitude)
        let destination = CLLocation(latitude: latitude, longitude: longitude)
        return origin.distance(from: destination) / 1_000
    }

    static func formattedDistance(latitude: Double, longitude: Double, language: AppLanguage) -> String {
        let value = distanceInKilometres(latitude: latitude, longitude: longitude)
        let formatter = NumberFormatter()
        formatter.numberStyle = .decimal
        formatter.minimumFractionDigits = value < 10 ? 1 : 0
        formatter.maximumFractionDigits = value < 10 ? 1 : 0
        formatter.locale = Locale(identifier: language == .de ? "de_DE" : "en_GB")
        let number = formatter.string(from: NSNumber(value: value)) ?? String(format: "%.1f", value)
        return language == .de ? "\(number) km Luftlinie" : "\(number) km straight-line"
    }
}

File: web/src/lib/geo.ts
const GUESTHOUSE = { latitude: 49.4224662, longitude: 10.9807374 };

const radians = (degrees: number) => degrees * Math.PI / 180;

export function distanceInKilometres(latitude: number, longitude: number): number {
  const earthRadius = 6371;
  const latitudeDelta = radians(latitude - GUESTHOUSE.latitude);
  const longitudeDelta = radians(longitude - GUESTHOUSE.longitude);
  const a = Math.sin(latitudeDelta / 2) ** 2
    + Math.cos(radians(GUESTHOUSE.latitude)) * Math.cos(radians(latitude)) * Math.sin(longitudeDelta / 2) ** 2;
  return earthRadius * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
}

export function formattedDistance(latitude: number, longitude: number, locale: 'de' | 'en'): string {
  const distance = distanceInKilometres(latitude, longitude);
  const digits = distance < 10 ? 1 : 0;
  const number = new Intl.NumberFormat(locale === 'de' ? 'de-DE' : 'en-GB', {
    minimumFractionDigits: digits, maximumFractionDigits: digits,
  }).format(distance);
  return locale === 'de' ? `${number} km Luftlinie` : `${number} km straight-line`;
}
